"Enqueue paint timing" and "Enqueue event timing" steps should come after intersection observer updates

smfr opened the pull request:
#### 44555b62da739802bfb053c9e697acfb6ccc3846
<pre>
&quot;Enqueue paint timing&quot; and &quot;Enqueue event timing&quot; steps should come after intersection observer updates
<a href="https://bugs.webkit.org/show_bug.cgi?id=298783">https://bugs.webkit.org/show_bug.cgi?id=298783</a>
<a href="https://rdar.apple.com/160477586">rdar://160477586</a>

Reviewed by Tim Nguyen.

Looking at the combination of:

<a href="https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model">https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model</a>
<a href="https://w3c.github.io/event-timing/#sec-modifications-HTML">https://w3c.github.io/event-timing/#sec-modifications-HTML</a>
<a href="https://w3c.github.io/paint-timing/#sec-modifications-dom">https://w3c.github.io/paint-timing/#sec-modifications-dom</a>

The steps should be:

19. run the update intersection observations steps
20. record rendering time for doc given unsafeStyleAndLayoutStartTime.
21. mark Paint Timing for doc.
22. dispatch pending Event Timing entries

So introduce `PaintTiming` and `EventTiming` steps. It&apos;s necessary to keep these steps in
`Page::doAfterUpdateRendering()` because they have to happen after layout, since LCP needs
to query geometry.

Tested by WTP.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/300531@main">https://commits.webkit.org/300531@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d2cc6793a99291b58aa92c54135b16c92d0564ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42520 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129434 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74912 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/41bb1db0-c0e3-4b52-881d-62173fb065f8) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124684 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43359 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93334 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61958 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125760 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34578 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109932 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73974 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2acb073e-48b4-4aa7-b525-bd2136689d62) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33558 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28089 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72926 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104288 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28302 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132162 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49753 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37876 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101933 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106148 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101799 "Found 3 new API test failures: TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25866 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47091 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25275 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46524 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49611 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55363 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49077 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52429 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50760 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d2cc6793a99291b58aa92c54135b16c92d0564ef. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50682

smfr committed:
"Enqueue paint timing" and "Enqueue event timing" steps should come after intersection observer updates
https://bugs.webkit.org/show_bug.cgi?id=298783
rdar://160477586

Reviewed by Tim Nguyen.

Looking at the combination of:

https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-processing-model
https://w3c.github.io/event-timing/#sec-modifications-HTML
https://w3c.github.io/paint-timing/#sec-modifications-dom

The steps should be:

19. run the update intersection observations steps
20. record rendering time for doc given unsafeStyleAndLayoutStartTime.
21. mark Paint Timing for doc.
22. dispatch pending Event Timing entries

So introduce `PaintTiming` and `EventTiming` steps. It's necessary to keep these steps in
`Page::doAfterUpdateRendering()` because they have to happen after layout, since LCP needs
to query geometry.

Tested by WTP.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
(WebCore::operator<<):
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/300531@main
